python/mlc_chat/model/llama/llama_model.py

```diff
@@ -228,48 +293,153 @@ def _attention_mask(batch_size, seq_len, total_seq_len):
-        return op.softmax(logits / temperature, axis=-1)
+        return op.softmax(logits / op.reshape(temperature, (temperature.shape[0], 1, 1)), axis=-1)
+
+    def create_paged_kv_cache(
+        self, max_batch_size: tir.Var, max_total_seq_len: tir.Var, page_size: tir.Var
+    ) -> PagedKVCache:
+        import tvm  # pylint: disable=import-outside-toplevel
+        from tvm import relax  # pylint: disable=import-outside-toplevel
+
+        from ..tir_inventory import (  # pylint: disable=import-outside-toplevel
+            kv_cache_debug_get_kv,
+            kv_cache_transpose_append,
+        )
+
+        attn_prefill = tvm.get_global_func(
+            "paged_kv_cache.attention_kernel_prefill", allow_missing=True
+        )
+        if attn_prefill is None:
+            return op.zeros((), self.dtype)
+
+        bb = relax.BlockBuilder.current()
+        num_qo_heads = self.num_attention_heads // self.tensor_parallel_shards
+        num_kv_heads = self.num_key_value_heads // self.tensor_parallel_shards
+        return PagedKVCache.create(
+            max_batch_size=max_batch_size,
+            max_total_seq_len=max_total_seq_len,
+            page_size=page_size,
+            num_hidden_layers=self.num_hidden_layers,
+            num_attention_heads=num_qo_heads,
+            num_key_value_heads=num_kv_heads,
+            head_dim=self.head_dim,
+            rope_scale=1,
+            rope_theta=self.rope_theta,
+            dtype=self.dtype,
+            f_transpose_append=bb.add_func(
+                kv_cache_transpose_append(num_kv_heads, self.head_dim, self.dtype),
+                "kv_cache_transpose_append",
+            ),
+            f_attn_prefill="paged_kv_cache.attention_kernel_prefill",
+            f_attn_decode="paged_kv_cache.attention_kernel_decode",
+            f_attn_prefill_ragged="flashinfer.attention_kernel_prefill_with_ragged_kv_cache",
+            f_attn_prefill_begin_forward="paged_kv_cache.attention_kernel_prefill_begin_forward",
+            f_attn_prefill_end_forward="paged_kv_cache.attention_kernel_prefill_end_forward",
+            f_attn_decode_begin_forward="paged_kv_cache.attention_kernel_decode_begin_forward",
+            f_attn_decode_end_forward="paged_kv_cache.attention_kernel_decode_end_forward",
+            f_attn_prefill_ragged_begin_forward="flashinfer.attention_kernel_prefill_with_ragged_kv_cache_begin_forward",  # pylint: disable=line-too-long
+            f_attn_prefill_ragged_end_forward="flashinfer.attention_kernel_prefill_with_ragged_kv_cache_end_forward",  # pylint: disable=line-too-long
+            f_attn_merge_state="flashinfer.merge_state_in_place",
+            f_attn_apply_rope="flashinfer.batch_qk_apply_rotary_in_place",
+            f_debug_get_kv=bb.add_func(
+                kv_cache_debug_get_kv(
+                    self.num_hidden_layers, num_kv_heads, self.head_dim, self.dtype
+                ),
+                "kv_cache_debug_get_kv",
+            ),
+        )
```

Comment:
i'd love to suggest if we could organize this piece a bit to avoid potential confusion. There are a couple of things we could potentially do:
- Extend (or add a new API that mimics) `nn.tensor_ir_op` to allow adding TIR/TE as operators
- Default methods like the one below are unlikely necessary and please remove

```
f_attn_prefill="paged_kv_cache.attention_kernel_prefill",
```

- Checking if a global func exists in the compiler (line #345) doesn't necessarily related to whether or not it exists in runtime, and we might want to think of other alternatives.

A rule of thumb is that we don't want to leak "tvm.relax" anywhere other than the `compiler_passes` package unless it is really really necessary.

Reply:
Most of the function below can be moved to model/kv_cahce.py, 

Add a subclass FlashInferKVCache that subclasses KVCache and supply all the related functions. The block builder can also be obtained there. In this case, we make kv_cache.py aware of the relax but the upper layers won't

Reply:
> * Extend (or add a new API that mimics) `nn.tensor_ir_op` to allow adding TIR/TE as operators

One need here is that “we want to fetch the GlobalVar after inserting a TIR function to IRModule,” where the GlobalVar is passed to KV cache constructor later on. Does it look good to add a new op or extending an existing one that returns GlobalVar, which is an IR construct that does not often exposed in the most-front ends?

Reply:
Okay I hided the GlobalVar and BlockBuilder logic to `kv_cache.py` with subclassing. Please take another look! Right now in `llama_model.py` it just simply 
```python
def create_paged_kv_cache(
    self, max_batch_size: tir.Var, max_total_seq_len: tir.Var, page_size: tir.Var
) -> PagedKVCache:
    num_qo_heads = self.num_attention_heads // self.tensor_parallel_shards
    num_kv_heads = self.num_key_value_heads // self.tensor_parallel_shards
    # Note: Right now we only have FlashInfer-based KV cache supported.
    # TIR version will be introduced soon.
    return FlashInferPagedKVCache.create(
        max_batch_size=max_batch_size,
        max_total_seq_len=max_total_seq_len,
        page_size=page_size,
        num_hidden_layers=self.num_hidden_layers,
        num_attention_heads=num_qo_heads,
        num_key_value_heads=num_kv_heads,
        head_dim=self.head_dim,
        rope_scale=1,
        rope_theta=self.rope_theta,
        dtype=self.dtype,
    )
```

Reply:
> * Checking if a global func exists in the compiler (Line 345) doesn't necessarily related to whether or not it exists in runtime, and we might want to think of other alternatives.

Thanks for pointing out. I removed the check so that compilation generates good `create_paged_kv_cache`. Right now it will error out at runtime in VM when the particular PackedFuncs do not exist.